node/actors/network/src/gossip/attestation/mod.rs

```diff
@@ -0,0 +1,330 @@
+        Ok(())
+    }
+
+    fn insert_votes(
+        &mut self,
+        votes: impl Iterator<Item = Arc<attester::Signed<attester::Batch>>>,
+    ) -> anyhow::Result<()> {
+        let mut done = HashSet::new();
+        for vote in votes {
+            // Disallow multiple entries for the same key:
+            // it is important because a malicious attester may spam us with
+            // new versions and verifying signatures is expensive.
+            if done.contains(&vote.key) {
+                anyhow::bail!("duplicate entry for {:?}", vote.key);
+            }
+            done.insert(vote.key.clone());
+            self.insert_vote(vote)?;
+        }
+        Ok(())
+    }
+
+    fn qc(&self) -> Option<attester::BatchQC> {
+        if self.weight < self.config.committee.threshold() {
+            return None;
+        }
+        let mut sigs = attester::MultiSig::default();
+        for vote in self.votes.values() {
+            sigs.add(vote.key.clone(), vote.sig.clone());
+        }
+        Some(attester::BatchQC {
+            message: self.config.batch_to_attest.clone(),
+            signatures: sigs,
+        })
+    }
+}
+
+/// Receiver of state diffs.
+pub(crate) struct DiffReceiver {
+    prev: Option<State>,
+    recv: sync::watch::Receiver<Option<State>>,
+}
+
+impl DiffReceiver {
+    /// Waits for the next state diff.
+    pub(crate) async fn wait_for_diff(&mut self, ctx: &ctx::Ctx) -> ctx::OrCanceled<Diff> {
+        loop {
+            let Some(new) = (*sync::changed(ctx, &mut self.recv).await?).clone() else {
+                continue;
+            };
+            let diff = new.diff(&self.prev);
+            self.prev = Some(new);
+            if !diff.is_empty() {
+                return Ok(diff);
+            }
+        }
+    }
+}
+
+/// `Controller` manages the attestation state.
+/// It maintains a set of votes matching the attestation config.
+/// It allows for
+/// * adding votes to the state
+/// * subscribing to the vote set changes
+/// * waiting for the certificate to be collected
+///
+/// It also keeps an attester key used to sign the batch vote,
+/// whenever it belongs the current attester committee.
+/// Signing happens automatically whenever the committee is updated.
+///
+/// Expected usage:
+/// ```
+/// let ctrl = attestation::Controller::new(Some(key));
+/// loop {
+///     // Check the global attestation registry.
+///     // Compute the next expected batch and the committee that should attest it.
+///     ...
+///     let config = attestation::Config {
+///         batch_to_attest: ...,
+///         committee: ...,
+///     };
+///     ctrl.update_config(Arc::new(config.clone())).unwrap();
+///     s.spawn(async {
+///         if let Some(qc) = ctrl.wait_for_qc(ctx, config.batch_to_attest.number).await?;
+///         // Submit the certificate `qc` to the global registry
+///         ...
+///     });
+///     // Wait for the global registry to include the certificate.
+///     ...
+/// }
```

Comment:
This example looks deceptively simple. 

What looked suspicious to me is that `wait_for_qc` might never complete if we miss the votes through gossip _and_ the `config.batch_to_attest.number` never changes to reflect the fact that the main node API now indicates that we should move on to vote on the next batch. Then I noticed the `s.spawn(...)` which makes this run in the background, so the main loop can move on. I assume `Wait for the global registry to include the certificate.` is supposed to prevent a tight loop, but it's not a given that every EN will have the certificate.

Reply:
Perhaps two concurrent loops would be simpler to understand:
* one to get the next batch number to attest, wait for the batch, wait for the committee, update the status
* another that just keeps calling `wait_for_qc` with a batch it increments by 1 regardless of the status and saves whever it gets `Some`

Reply:
>  but it's not a given that every EN will have the certificate.

By global registry, I mean L1/main node. LMK if you have a better word for this.

Reply:
I can make it 2 loops, yes.

Reply:
I think for the waiting part it could just say L1.

Reply:
> As discussed on the daily, Roman doesn't think waiting for the batch N to be on L1 before starting to collect votes for batch N+1 is feasible.

That won't be enough, because the current implementation doesn't support gossiping certificates for multiple batches in parallel (this doesn't change from the previous implementation), while we need to keep gossiping votes for batch N until it is actually committed to L1. We will address it later though.

Reply:
I might be missing the obvious, but I thought the issue was that we didn't want to wait for N to be on L1 before we start gossiping about N+1. If we only wait until we have the certificate for N in the `l1_batches_consensus` table, then we can immediately start gossiping about N+1 as soon as the commitment is ready, regardless of whether it's submitted to L1 or not. At least this is the case when the main node is the only sender of the batches. You are right if we want anyone to be able to send the data then we have to gossip until the (unknown) sender(s) manage to get the certificate, which will require keeping the votes around until the batch is on L1.

Reply:
That will work as long as we use the main node as the "global registry". When L1 is the global registry, each node should wait until the batch commitment is finalized on L1 before ceasing to gossip the attestations for that batch. But I don't know how committing batches in parallel will work in a multi-submitter setup though.